### PR TITLE
8310685: [lw5] adding support for null restricted types

### PR DESCRIPTION
**Pull request by vicente-romero-oracle**
initial support for null restricted types

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Issue
 * [JDK-8310685](https://bugs.openjdk.org/browse/JDK-8310685): [lw5] adding support for null restricted types (**Bug** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/valhalla.git pull/872/head:pull/872` \
`$ git checkout pull/872`

Update a local copy of the PR: \
`$ git checkout pull/872` \
`$ git pull https://git.openjdk.org/valhalla.git pull/872/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 872`

View PR using the GUI difftool: \
`$ git pr show -t 872`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/valhalla/pull/872.diff">https://git.openjdk.org/valhalla/pull/872.diff</a>

</details>

### CODE REVIEW

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java [hunk 1]
```diff
@@ -1136,10 +1163,17 @@ public Boolean visitType(Type t, Type s) {
                      return t.hasTag(s.getTag());
                  case TYPEVAR:
                      return isSubtypeNoCapture(t.getUpperBound(), s);
-                 case BOT:
+                 case BOT: {
+                     /* this method can be invoked even from the backend, and warnings can be printed again, so
+                      * make sure that the caller really wants to warn
+                      */
+                     if (s.isNonNullable() && warnStack.head.pos() != null) {
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java:1170]: This is a bit odd. We are defining a type relationship - if "BOT" is not a subtype of T! that is fine, we should just return false. Giving an error seems a bit strange here (I don't think it is normal for subtyping to generate errors).

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java [hunk 2]
```diff
@@ -1204,11 +1238,15 @@ public Boolean visitClassType(ClassType t, Type s) {
                 if (sup == null) return false;
                 // If t is an intersection, sup might not be a class type
                 if (!sup.hasTag(CLASS)) return isSubtypeNoCapture(sup, s);
-                return sup.tsym == s.tsym
+                boolean result = sup.tsym == s.tsym
                      // Check type variable containment
                     && (!s.isParameterized() || containsTypeRecursive(s, sup))
                     && isSubtypeNoCapture(sup.getEnclosingType(),
                                           s.getEnclosingType());
+                if (result && warnStack.head.pos() != null && s.hasNarrowerNullabilityThan(t)) {
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java:1246]: I have memories of the previous implementation where we discussed maybe keeping subtyping routines as is, but adding some other check to validate the "polarity" of the type arguments. Wouldn't something similar work here too?

File: src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/Gen.java [hunk 3]
```diff
@@ -2115,6 +2123,10 @@ public void visitParens(JCParens tree) {
     public void visitAssign(JCAssign tree) {
         Item l = genExpr(tree.lhs, tree.lhs.type);
         genExpr(tree.rhs, tree.lhs.type).load();
+        if (tree.lhs.type.isNonNullable()) {
+            code.emitop0(dup);
+            genNullCheck(tree.rhs);
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/Gen.java:2128]: Should the cast always be generated even if we know that the RHS is non-nullable? (I'm not sure here)

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java (hunk 1 above)
Comment (vicente-romero-oracle) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java:1170]: yep you are right

**Pull request closed without merging**